proposals/state-machine-restriction-updates.md

```diff
@@ -0,0 +1,167 @@
+# State machine restriction updates
+
+## Summary
+[summary]: #summary
+
+- Allow `ref`/`ref struct` locals and `unsafe` blocks in iterators and async methods
+  provided they are used in code segments without any `yield` or `await`.
+- Warn about `yield` inside `lock`.
+- Consolidate spec: disallow pointers to hoisted variables
+  (pointers to captured variables are already disallowed).
+
+## Motivation
+[motivation]: #motivation
+
+It is not necessary to disallow `ref`/`ref struct` locals and `unsafe` blocks in async/iterator methods
+if they are not used across `yield` or `await`, because they do not need to be hoisted.
+
+```cs
+async void M()
+{
+    await ...;
+    ref int x = ...; // error previously, proposed to be allowed
+    x.ToString();
+    await ...;
+    // x.ToString(); // still error
+}
+```
+
+On the other hand, having `yield` inside a `lock` means the caller also holds the lock while iterating which might lead to unexpected behavior.
+This is even more problematic in async iterators where the caller can `await` between iterations, but `await` is not allowed in `lock`.
+See also https://github.com/dotnet/roslyn/issues/72443.
+
+```cs
+lock (this)
+{
+    yield return 1; // warning proposed
+}
+```
+
+## Detailed design
+[design]: #detailed-design
+
+[§13.3.1 Blocks > General][blocks-general]:
+
+> ~~It is a compile-time error for an iterator block to contain an unsafe context ([§23.2][unsafe-contexts]).~~
+> An iterator block always defines a safe context, even when its declaration is nested in an unsafe context.
+
+[§13.6.2.4 Ref local variable declarations][ref-local]:
+
+> ~~It is a compile-time error to declare a ref local variable, or a variable of a `ref struct` type,
+> within a method declared with the *method_modifier* `async`, or within an iterator ([§15.14][iterators]).~~
+> **It is a compile-time error to declare and use a ref local variable, or a variable of a `ref struct` type
+> across `await` or `yield` statements.**
+
+[§13.13 The lock statement][lock-statement]:
+
+> [...]
+> 
+> **A warning is reported (as part of the next warning wave) when a `yield` statement
```

Comment:
Is this only for legacy `lock`? I assume it will be an error to yield out of a "modern" lock.

Reply:
Yes, the modern lock cannot work with `yield` since it's lowered to `using` of a `ref struct`. I will make it clearer.